### DISCUSSION TITLE
Init multiple indexes and store them

### DISCUSSION THREAD

**Discussion by Rigo-m** [nuxt-modules/algolia#24] [Ideas]
Shouldn't `useInitIndex` be more capable and init more indexes at once and maybe save them into a global state accessible through the app? Could be useful when working with a big application with lots of ui components working with indexes.

**Comment by Baroshem** [comment]
Hey, 

If I understand you correctly, you would like `useInitIndex` to accept a string or an array of strings (of index names)?
And then use `useState` with some dynamic value to be able to store multiple indexes in both client and server reactive state?

Something like this:

```vue
<script setup>
  const indexes = useInitIndex(['test', 'super', 'etc'])

  indexes[0].appId = 'test'
</script>
```

This is indeed interesting feature, however I am not sure if this should be part of this composable or whether it should be a completely new composable like `useInitMultipleIndexes`. Thanks to that we will have one composable that handles single index and other that handles many indexes. We could avoid issues with different composable params and return values.

What do you think?

**Reply by Rigo-m** [reply]
I'll elaborate a bit:
I have an application that has multiple indexes right now, e.g.: `products`, `models`, `product_families`,  `news`, `stories` etc...
What I do, is that I export the indices from a js file like so:
```js
import algoliasearch from 'algoliasearch'
export const algoliaClient = algoliasearch(
  process.env.ALGOLIA_APP_ID,
  process.env.ALGOLIA_SEARCH_API_KEY,
)

export default {
  client: algoliaClient,
  indices: {
    products: algoliaClient.initIndex('MY_LONG_INDEX_NAME_FOR_PRODUCTS'),
    models: algoliaClient.initIndex('MY_MODEL_INDEX')
    // etc
  }
}
```

And then I can import in any composable this:
```js
import algolia from '~/config/algolia'

export const useMyComposable = () => {
  const index = algolia.indices.products
}
```

As you can see this approach is not typed, but you got my point. I can init the indexes in one place, and handily import them in many places.

What do you think about keeping `useInitIndex` for simple, single/not too many indexes applications, and also propose another approach for more complex apps?

For example, we could use the module configuration to accept a series of indexes and then expose them in the plugin. The API would look something like this:

`nuxt.config.js`
```js
algolia: {
  indexes: {
    alias: 'MY_INDEX_NAME_TO_ALIAS`,
    other_index: 'MY_OTHER_INDEX`
  }
}
```

Then, somewhere in my code:
```js
export const myCustomComposable = () => {
  // i could use useSearch
  const { result, search } = useSearch('products')
  // or fine-grain my composable by importing and using multiple aliases
  const { $algolia } = useNuxtApp()
  const { alias, other_index } = $algolia.indices
  // do something with my indexes ...
}
```

**Reply by Baroshem** [reply]
I see, so it's more like to store the names of the indexes in the configuration file rather than initializing them using algoliasearch?

Something like:

```ts
// nuxt.config.js

{
    buildModules: [
        "@nuxtjs/algolia",
    ],
    algolia: {
        apiKey: "<YOUR_SEARCH_API_KEY>",
        applicationId: "<YOUR_APPLICATION_ID>",
        indexes: {
           main_index: 'NAME_OF_MY_INDEX'
        }
    }
}
```

And then use it like this:

```vue
// page.vue
<script setup>
const { algolia } = useRuntimeConfig()   // <- I think you will have access here instead of NuxtApp as NuxtApp will contain reference to algoliasearch while config will contain your configuration 
const { result, search } = useSearch(algolia.indexes.main_index);

onMounted(async () => {
  await search({ query: "Samsung" });
});
</script>
```

Sorry if I am repeating your words, just want to know that we are on the same page as I like the idea :)

**Reply by Rigo-m** [reply]
Exactly this + a composable to init multiple indices at once in case we need to work with multiple indexes at once.

i.e.: 
```js
const { algolia } = useRuntimeConfig()
const { productIndex, moduleIndex } = useInitMultipleIndex(algolia.indexes.products, algolia.indexes.models)`
```

**Reply by Rigo-m** [reply]
I think having indexes defined in one place and create types from them can greatly improve DX by suggesting the correct indexes names and avoid trivial errors.

**Reply by Baroshem** [reply]
That would work really well. I would just change it so that the `useInitMultipleIndex()` composable would accept an array and also return an array of indexes like this:

```ts
const { algolia } = useRuntimeConfig()
const indexes = useInitMultipleIndex([algolia.indexes.products, algolia.indexes.models])

console.log("product-index", indexes[0])
console.log("models-index", indexes[1])
```

**Reply by Rigo-m** [reply]
Mind me making a PR to make this happen?

**Reply by Baroshem** [reply]
That would be awesome! Yes, do it for sure and I will do my best to help you :)

**Comment by Baroshem** [comment]
@yassilah 

What do you think about this feature? :)

**Comment by yassilah** [comment]
Hmm I see the point of storing the index names somewhere but I don't really see the value of having it as part the module. What I think could be interesting though is to provide and extensible interface to add types and IDE auto-complete for index names. This way we don't need to store the index names but they would still be "configurable" and add a lot more value (in my opinion) as you would get search results types depending on the index name. I think initializing multiple indices can be a good thing though and shouldn't be too hard to implement :)

**Reply by Rigo-m** [reply]
Is it possible to somehow type algolia responses? Contentful, for example, has a CLI that download typescripts content types to the client

**Reply by yassilah** [reply]
Yes,  it's possible but those would just be the **expected** result types that you would type yourself. Something simple like:

```ts 
declare module '@nuxtjs/algolia' {
  interface AlgoliaIndices {
    foo: {
      bar1: string
      bar2: number
    }
  }
}
```

and then:
```vue
<script lang="ts" setup>
const { search } = useInitIndex('foo') // Index name auto-completion

const { hits } = await search('something')

hits[0].bar1 // string
hits[0].bar2 // number
</script>
```

You can already do so right now by providing the type when searching `index.search<Type>('something')` but here it would do that automatically based on the provided index name.

**Reply by Baroshem** [reply]
In this case, I would stick with typing the index.search. Not sure though if search method accepts a generic type that can be then passed to index.search method. Maybe that would be more useful?

**Reply by yassilah** [reply]
I think we're talking about the same thing but I'm not sure aha I'll just make a quick example on another branch, you may take a look and tell me if that's it, if not just discard it 👍

**Reply by yassilah** [reply]
there you go: [feat/type-indices](https://github.com/nuxt-community/algolia-module/tree/feat/type-indices)

**Reply by Baroshem** [reply]
Yeah, this is a good example however, all of this can be done already as you added all changes to playground (what also makes me wanna have this as a part of algolia docs ;)). Would you like to create a short section in the docs about typing indices example? That could be useful for someone in the future.

Regarding the multiple indexes. I tried to work with it for a bit and I found it a bit like overdoing simple things as this all can be done right now without additional composable.

**Reply by yassilah** [reply]
Hey :) No currently you can't do it, I had to add some typings magic into the composable and the new `types.ts` file. But if we want to add those, I'd be happy to write some docs

**Reply by Baroshem** [reply]
You are right. I looked at only one commit. Let's create a pull request of your changes and see how it works in battle :)

**Comment by Baroshem** [comment]
Hey @Rigo-m 

Yassilah created a pull request with this naming and type completion indices functionality. Could you take a look at it and give your thoughts? :) 

https://github.com/nuxt-community/algolia-module/pull/29

**Comment by Baroshem** [comment]
Guys,

I am really happy to announce that this feature was merged and released in a new version 0.7.1 :)

**Reply by Baroshem** [reply]
@Rigo-m 

Could you please mark the discussion as answered? :)
